### ISSUE TITLE
Improve GitHub actions

### ISSUE THREAD

**Issue by larson-carter** [ohmyzsh/ohmyzsh#8922]
**Is your feature request related to a particular plugin or theme? If so, specify it.**
No

**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Better GitHub actions, currently it just installs OhMyZsh. We need it to test themes for building, plugins, etc.

**Additional context**
- Actions
  - Optimize main action:
    * install only when changes made to install script
    * cache zsh installation
    * check only changed files
We don't need to install it every time. Only when there is a modification to the install script. Therefore we could only install it 1 time then have the install in cache then utilize the update tool to pull it forward. Then it would detect what themes have been modified and test them. Same for plugins detect what plugins have been modified then test the plugins aliases.

Overall we need our GitHub actions to me thorough

**Comment by magdapoppins** [comment]
I could take a stab at this and maybe also add [CodeQL](https://securitylab.github.com/tools/codeql) as a part of actions if that seems useful?

**Comment by larson-carter** [comment]
@magdapoppins 

Sure, go ahead and try whatever part that you want.

> Also, could you elaborate on what "cache zsh installation" means and why this would be a good idea?

This might be impossible with GitHub actions I honestly don't know. The "cache zsh installation" was meaning save the zsh installation so that way it won't have to install it every time we run a workflow.

**Comment by magdapoppins** [comment]
Some thoughts after some initial stab:
- seeing as the current install process only takes about 2s, I don't see that there is a big need for optimization
- however, the current installation test only tests that master can be installed
  - this should be changed to test the branch from which the PR is made (relatively straightforward)
  - maybe the CI process could "install" OhMyZsh by symlinking the working copy into place and overwriting `.zshrc`
- perhaps some smoke tests could also be added: what do you think would be a good smoke test for ohmyzsh?

**Comment by mcornella** [comment]
> however, the current installation test only tests that master can be installed
> * this should be changed to test the branch from which the PR is made (relatively straightforward)

This is not true, the `install.sh` file used is the one from the PR head. What's true is that the *cloned* repository is the main repo, but that has no effect on whether the install script works or not.

> * seeing as the current install process only takes about 2s, I don't see that there is a big need for optimization

This however would be nice for the relatively few times when the [`apt install` process fails](https://github.com/ohmyzsh/ohmyzsh/pull/8961#issuecomment-654280533) for reasons beyond our control. We could use an external Docker image ([`zshusers/zsh`](https://hub.docker.com/r/zshusers/zsh), for example, but it doesn't have git installed).

> * maybe the CI process could "install" OhMyZsh by symlinking the working copy into place and overwriting `.zshrc`

This would be the best way to do it if we had tests to run and the PR didn't modify the install script.

> * perhaps some smoke tests could also be added: what do you think would be a good smoke test for ohmyzsh?

🤷🏻

**Comment by vijayb-aiops** [comment]
Hi, I’d like to work on this enhancement. Could you please assign it to me?

**Comment by robbyrussell** [comment]
Hey team, this particular issue feels a bit vague and given that we don't entirely have the slowest build, let's opt to open up individual Issues/PRs for incremental changes so that we're not chasing "improve"
